File: service_caller.go
package spine

import (
	"context"
	"fmt"
	"time"

	"github.com/cenkalti/backoff/v4"
	"github.com/poisnoir/mad-go"
	"github.com/poisnoir/spine-go/internal/globals"
	"github.com/xtaci/kcp-go/v5"
)

type ServiceCaller[K any, V any] struct {
	namespace    *Namespace
	serviceName  string
	keyEncoder   *mad.Mad[K]
	valueEncoder *mad.Mad[V]
	errorEncoder *mad.Mad[string]
	conn         *kcp.UDPSession
	ctx          context.Context
	requests     chan serviceRequest[K, V]
	isConnected  bool
	cancel       context.CancelFunc
}

func NewServiceCaller[K any, V any](namespace *Namespace, serviceName string) (*ServiceCaller[K, V], error) {

	keyEnc, err := mad.NewMad[K]()
	if err != nil {
		return nil, err
	}

	valueEnc, err := mad.NewMad[V]()
	if err != nil {
		return nil, err
	}

	errEnc, _ := mad.NewMad[string]()

	ctx, cancel := context.WithCancel(namespace.ctx)

	sc := &ServiceCaller[K, V]{
		namespace:    namespace,
		keyEncoder:   keyEnc,
		valueEncoder: valueEnc,
		serviceName:  serviceName,
		ctx:          ctx,
		cancel:       cancel,
		errorEncoder: errEnc,
		isConnected:  false,
		requests:     make(chan serviceRequest[K, V], 100),
	}

	go sc.run()

	return sc, nil
}

func (sc *ServiceCaller[K, V]) run() {

	// timer heartbeat
	// make sure connection is open and alive
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		if sc.isConnected {
			select {
			case <-ticker.C:
				if err := ping(sc.conn); err != nil {
					sc.isConnected = false
				}
			case requestData := <-sc.requests:
				// Postpone heartbeat
				ticker.Reset(10 * time.Second)
				output, err := sc.send(requestData.input)
				if err != nil {
					sc.isConnected = false
				}
				requestData.output <- serviceOutput[V]{data: output, err: err}
			// Todo: close the service caller
			case <-sc.ctx.Done():
				return
			}
		} else {
			bo := backoff.WithContext(backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(0)), sc.ctx)
			_ = backoff.Retry(sc.connect, bo)
		}
	}
}

// send is the gateway to kcp connection. It acts as multiplexer
func (sc *ServiceCaller[K, V]) send(key K) (V, error) {
	var v V

	requestSize := sc.keyEncoder.GetRequiredSize(&key) + 1
	if requestSize > globals.MAX_PACKET_SIZE {
		return v, fmt.Errorf(globals.ERROR_PAYLOAD_SIZE)
	}

	bufPtr := sc.namespace.bufferPool.Get().(*[]byte)
	defer sc.namespace.bufferPool.Put(bufPtr)
	buf := *bufPtr
	buf[0] = globals.SERVICE_REQUEST
	sc.keyEncoder.Encode(&key, buf[1:])

	_, err := write(sc.conn, buf, requestSize, true)
	if err != nil {
		return v, err
	}

	// Todo make different errors.
	// for now assume err is handler error
	if buf[0] != globals.OK_STATUS_CODE {
		return v, fmt.Errorf(globals.ERROR_SERVICE_HANDLER)
	}

	_ = sc.valueEncoder.Decode(buf, &v)
	return v, nil
}

// Call sends key to the service and returns V from service
// context is used for establishing connection
func (sc *ServiceCaller[K, V]) Call(key K, ctx context.Context) (V, error) {

	var zero V
	data := serviceRequest[K, V]{
		input:  key,
		output: make(chan serviceOutput[V], 1),
	}
	sc.requests <- data

	select {
	case <-ctx.Done():
		return zero, ctx.Err()
	case output := <-data.output:
		return output.data, output.err
	}
}

func (sc *ServiceCaller[K, V]) Close() {
	sc.cancel()
}

func (sc *ServiceCaller[K, V]) connect() error {

	logger := sc.namespace.logger.With(
		sc.namespace.Name(),
		"service_caller",
		sc.serviceName,
		"connect",
	)

	// finding the service
	address, err := sc.namespace.GetService(sc.serviceName, sc.ctx)
	if err != nil {
		logger.Error("unable to find the service", "error", err)
		return err // the only way to fail here is to run out of context
	}

	// establishing connection
	sess, err := kcp.DialWithOptions(address, sc.namespace.encryption, 10, 3)
	if err != nil {
		logger.Error("failed to dial service", "error", err)
		return err
	}

	// getting buffer for comm
	bufPtr := sc.namespace.bufferPool.Get().(*[]byte)
	defer sc.namespace.bufferPool.Put(bufPtr)
	buf := *bufPtr

	// validating input/output service types
	keyCode := sc.keyEncoder.Code()
	n := copy(buf, keyCode)

	n, err = write(sess, buf, n, true)
	if err != nil {
		logger.Error("failed to validate service input type", "error", err)
		return err
	} else if n != 1 {
		err = fmt.Errorf("response is corrupted")
		logger.Error("failed to validate service input type", "error", err)
		return err
	} else if buf[0] != globals.OK_STATUS_CODE {
		err = fmt.Errorf("service data type is different")
		logger.Error("failed to validate service input type", "error", err)
		return err
	}

	valueCode := sc.valueEncoder.Code()
	n = copy(buf, valueCode)

	n, err = write(sess, buf, n, true)
	if err != nil {
		logger.Error("failed to validate service output type", "error", err)
	} else if n != 1 {
		err = fmt.Errorf("response is corrupted")
		logger.Error("failed to validate service output type", "error", err)
		return err
	} else if buf[0] != globals.OK_STATUS_CODE {
		err = fmt.Errorf("service data type is different")
		logger.Error("failed to validate service output type", "error", err)
		return err
	}
	sc.conn = sess
	sc.isConnected = true

	return nil

}

File: py/service_caller.py
import asyncio
import logging
import time
from .mad import new_mad
from . import globals
from .network import ping

class ServiceCaller:
    def __init__(self, namespace, service_name):
        self.namespace = namespace
        self.service_name = service_name
        self.logger = namespace.logger.getChild(f"service_caller.{service_name}")
        
        self.key_encoder = new_mad()
        self.value_encoder = new_mad()
        self.error_encoder = new_mad("string")
        
        self.reader = None
        self.writer = None
        self.is_connected = False
        self.requests = asyncio.Queue(maxsize=100)
        self.ctx = asyncio.Event()
        self.tasks = []

    async def start(self):
        self.tasks.append(asyncio.create_task(self.run()))

    async def run(self):
        while not self.ctx.is_set():
            if self.is_connected:
                try:
                    # Timer for heartbeat (simplified)
                    # In a real implementation, we'd use wait_for with a timeout
                    try:
                        request_data, output_queue = await asyncio.wait_for(self.requests.get(), timeout=10.0)
                        
                        # Send request
                        try:
                            result = await self.send(request_data)
                            await output_queue.put((result, None))
                        except Exception as e:
                            self.logger.error(f"Send failed: {e}")
                            self.is_connected = False
                            await output_queue.put((None, e))
                        finally:
                            self.requests.task_done()
                            
                    except asyncio.TimeoutError:
                        # Ping
                        try:
                            await ping(self.reader, self.writer)
                        except Exception as e:
                            self.logger.error(f"Ping failed: {e}")
                            self.is_connected = False
                except asyncio.CancelledError:
                    break
                except Exception as e:
                    self.logger.error(f"Run loop error: {e}")
                    self.is_connected = False
            else:
                # Try to connect with exponential backoff
                await self.connect_with_retry()

    async def connect_with_retry(self):
        delay = 1.0
        while not self.is_connected and not self.ctx.is_set():
            try:
                await self.connect()
                self.is_connected = True
                self.logger.info(f"Connected to {self.service_name}")
            except Exception as e:
                self.logger.error(f"Failed to connect to {self.service_name}: {e}")
                await asyncio.sleep(delay)
                delay = min(delay * 2, 60.0)

    async def connect(self):
        # Finding the service
        addr = await self.namespace.get_service_addr(self.service_name)
        host, port = addr.split(':')
        port = int(port)

        # Establishing KCP connection
        # Placeholder for actual KCP library connection
        # self.reader, self.writer = await kcp.open_connection(host, port)
        
        # Simulating handshake for now
        # writer.write(self.key_encoder.code())
        # ... validation ...
        # writer.write(self.value_encoder.code())
        # ... validation ...
        
        pass

    async def send(self, key):
        request_size = self.key_encoder.get_required_size(key) + 1
        if request_size > globals.MAX_PACKET_SIZE:
            raise Exception(globals.ERROR_PAYLOAD_SIZE)

        # Build packet: [OP_CODE][ENCODED_KEY]
        packet = bytes([globals.SERVICE_REQUEST]) + self.key_encoder.encode(key)
        
        self.writer.write(packet)
        # Wait for response
        response = await self.reader.read(globals.MAX_PACKET_SIZE)
        
        if not response:
            raise Exception("No response from service")
            
        if response[0] != globals.OK_STATUS_CODE:
            raise Exception(globals.ERROR_SERVICE_HANDLER)

        return self.value_encoder.decode(response[1:])

    async def call(self, key, timeout=None):
        output_queue = asyncio.Queue(maxsize=1)
        await self.requests.put((key, output_queue))
        
        try:
            if timeout:
                return await asyncio.wait_for(output_queue.get(), timeout=timeout)
            else:
                return await output_queue.get()
        except asyncio.TimeoutError:
            raise Exception("Call timeout")

    async def close(self):
        self.ctx.set()
        for task in self.tasks:
            task.cancel()
        if self.writer:
            self.writer.close()
            # await self.writer.wait_closed()
